crates/wash/src/lib/config.rs

```diff
@@ -21,40 +21,326 @@ pub const DEFAULT_NATS_TIMEOUT_MS: u64 = 2_000;
+        }
+    }
+
+    fn runtime_dir(&self) -> Option<PathBuf> {
+        match self {
+            Self::Xdg(s) => s.runtime_dir(),
+            Self::Windows(s) => s.runtime_dir(),
+        }
+    }
+}
+
+macro_rules! in_dir_method {
+    ($self: ident, $path_extra: expr, $dir_method_name: ident) => {{
+        let mut path = $self.$dir_method_name();
+        path.push(Path::new(&$path_extra));
+        path
+    }};
+}
+
+macro_rules! create_dir_method {
+    ($self: ident, $dir_method_name: ident) => {{
+        let path = $self.$dir_method_name();
+        fs::create_dir_all(&path)
+            .with_context(|| format!("failed to create directory `{}`", path.display()))?;
+        Ok(path)
+    }};
+}
+
+macro_rules! create_in_dir_method {
+    ($self: ident, $path_extra: expr, $file_extra: expr, $dir_method_name: ident, $in_dir_method_name: ident) => {{
+        let mut path = if let Some(subdir) = $path_extra {
+            $self.$in_dir_method_name(subdir)
+        } else {
+            $self.$dir_method_name()
+        };
+        fs::create_dir_all(&path)
+            .with_context(|| format!("failed to create directory `{}`", path.display()))?;
+        if let Some(file) = $file_extra {
+            path.push(Path::new(&file));
+        }
+        Ok(path)
+    }};
+}
+
+impl WashAppStrategy {
+    /// Creates the directory in which to store configuration in and returns the path to it.
+    pub fn create_config_dir(&self) -> Result<PathBuf> {
+        create_dir_method!(self, config_dir)
+    }
+
+    /// Creates the directory in which to store configuration in and returns the path to it
+    /// potentially concatenated with a filename.
+    pub fn create_in_config_dir<P: AsRef<OsStr>>(
+        &self,
+        subdir: Option<P>,
+        file: Option<P>,
+    ) -> Result<PathBuf> {
+        create_in_dir_method!(self, subdir, file, config_dir, in_config_dir)
+    }
+
+    /// Returns the path to the directory where to store keys.
+    pub fn keys_dir(&self) -> PathBuf {
+        var_path("WASH_KEYS_DIR").unwrap_or_else(|_| self.in_data_dir("keys"))
+    }
+
+    /// Concatenates a path in the keys directory.
+    pub fn in_keys_dir<P: AsRef<std::ffi::OsStr>>(&self, path: P) -> PathBuf {
+        in_dir_method!(self, path, keys_dir)
+    }
+
+    /// Creates the directory in which to store keys in and returns the path to it.
+    pub fn create_keys_dir(&self) -> Result<PathBuf> {
+        create_dir_method!(self, keys_dir)
+    }
+
+    /// Creates the directory in which to store keys in and returns the path to it potentially
+    /// concatenated with a filename.
+    pub fn create_in_keys_dir<P: AsRef<OsStr>>(
+        &self,
+        subdir: Option<P>,
+        file: Option<P>,
+    ) -> Result<PathBuf> {
+        create_in_dir_method!(self, subdir, file, keys_dir, in_keys_dir)
+    }
+
+    /// Returns the path to the directory where to store logs.
+    pub fn logs_dir(&self) -> PathBuf {
```

Comment:
[ ] currently these functions are clearly not used everywhere, as the log directory for instance did not exist before. We will need to move the logs into this directory.

Reply:
No problem, I don't think we need to do any work to migrate old logs for example they are primarily throwaway 😄

Reply:
hahaha I did not mean migrations of old logs. But I think the `launch.log` is currently not being written in the new `logs` directory. I still need to figure out where that log initially originates and update the code. Also, the issue says it wants the directory to be `$WASH_DATA_DIR/logs/`. Per XDG logs should be under `$XDG_STATE_HOME`, and typically log directories are called `log` and not `logs`.

I would rather suggest that logs follow either of the following conventions:
- `$XDG_STATE_HOME/wash/launch.log` = `~/.local/state/wash/launch.log`
- `$XDG_STATE_HOME/wash/log/launch.log` = `~/.local/state/wash/log/launch.log`

Of course, one could keep the idea that `$WASH_LOGS_DIR` is `$XDG_STATE_HOME/wash[/log]` and use that. What do you think?

Reply:
Mhhh did not find any part in the code that actually writes that `launch.log` file. The only logs I can see with some primitive testing are the logs being generated by `wash dev` which are put into the `/dev` sub-directory which makes sense in my opinion.